core/lease/storedb.go

```diff
@@ -0,0 +1,365 @@
+// If either of these conditions is false, an error is returned.
+func (s *DBStore) RevokeLease(lease Key, holder string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+DELETE FROM lease
+WHERE  uuid = (
+    SELECT l.uuid
+    FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+    WHERE  t.type = ?
+    AND    l.model_uuid = ?
+    AND    l.name = ?
+    AND    l.holder = ?
+)`[1:]
+
+		result, err := s.db.ExecContext(
+			ctx, q, lease.Namespace, lease.ModelUUID, lease.Lease, holder)
+
+		if err == nil {
+			var affected int64
+			affected, err = result.RowsAffected()
+			if affected == 0 && err == nil {
+				err = ErrInvalid
+			}
+		}
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.New("revoke cancelled")
+	case err := <-errCh:
+		cancel()
+		return errors.Trace(err)
+	}
+}
+
+// LeaseGroup (lease.Store) returns all leases
+// for the input namespace and model.
+func (s *DBStore) LeaseGroup(namespace, modelUUID string) (map[Key]Info, error) {
+	q := `
+SELECT t.type, l.model_uuid, l.name, l.holder, l.expiry
+FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+WHERE  t.type = ?
+AND    l.model_uuid = ?`[1:]
+
+	rows, err := s.db.Query(q, namespace, modelUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result, err := leasesFromRows(rows)
+	return result, errors.Trace(err)
+}
+
+// PinLease (lease.Store) adds the input entity into the lease_pin table
+// to indicate that the lease indicated by the input key must not expire,
+// and that this entity requires such behaviour.
+func (s *DBStore) PinLease(lease Key, entity string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+INSERT INTO lease_pin (uuid, lease_uuid, entity_id)
+SELECT ?, l.uuid, ?
+FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+WHERE  t.type = ?
+AND    l.model_uuid = ?
+AND    l.name = ?`[1:]
+
+		_, err := s.db.ExecContext(
+			ctx, q, utils.MustNewUUID().String(), entity, lease.Namespace, lease.ModelUUID, lease.Lease)
+
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.New("pin lease cancelled")
+	case err := <-errCh:
+		cancel()
+		if isUniquenessViolation(err) {
+			return nil
+		}
+		return errors.Trace(err)
+	}
+}
+
+// UnpinLease (lease.Store) removes the record indicated by the input
+// key and entity from the lease pin table, indicating that the entity
+// no longer requires the lease to be pinned.
+// When there are no entities associated with a particular lease,
+// it is determined not to be pinned, and can expire normally.
+func (s *DBStore) UnpinLease(lease Key, entity string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+DELETE FROM lease_pin
```

Comment:
It would be an interesting exercise to think about what these complicated joins look like with sqlair. Since it is primarily input parameters, it should be relatively straightforward, I think. Though you do end up with a "uuid" that doesn't exist on your Key, and an entity that is just a string. So unless we craft a custom type for input, there isn't a simple named thing for all the parameters.

Reply:
There is a [UUID extension](https://sqlite.org/src/file/ext/misc/uuid.c) for SQLite, though I don't know the feasibility of compiling it in when building Dqlite.

It would be easy then to trigger if-null creation upon insert.

I would like to see us using UUID on our domain objects though - it would be great for traceability to pass them with requests *into* the API and have objects uniquely identified in the wider system whether-or-not they have made to the the point of persistence.